includes/blocks/course-theme/class-quiz-content.php

```diff
@@ -25,32 +25,39 @@ class Quiz_Content {
 	 */
 	public static function render_quiz() {
 
+		// The following content are rendered separately in Learning Mode.
+		// So we need to remove them from here, otherwise they are repeated.
 		remove_action( 'sensei_single_quiz_questions_before', [ Sensei()->post_types->messages, 'send_message_link' ], 10 );
 		remove_action( 'sensei_single_quiz_questions_after', [ 'Sensei_Quiz', 'action_buttons' ], 10 );
+		remove_action( 'sensei_single_quiz_content_inside_before', array( 'Sensei_Quiz', 'the_user_status_message' ), 40 );
+		remove_action( 'sensei_single_quiz_content_inside_before', array( 'Sensei_Quiz', 'the_title' ), 20 );
 
-		\Sensei_Quiz::start_quiz_questions_loop();
+		ob_start();
+
+		do_action( 'sensei_single_quiz_content_inside_before', get_the_ID() );
```

Comment:
That block does run most of the other hooks already, it's just a wrapper around the existing questions loop rendering via PHP templates. We do have [plans](https://github.com/Automattic/sensei/issues/4614) to move that to a more block-based approach, but I wonder if we still want to keep most of the quiz and question hooks, both for compatibility and extensibility. 
But the quiz save listener should definitely move out of any template/block filters, we can just add that to `init`/`template_redirect` like we did for lesson requests.

Reply:
I think it's best if we use the same [quiz template](https://github.com/automattic/sensei/blob/master/templates/single-quiz.php) that is used in regular mode to render the quiz in Learning Mode. So all the action hooks are the same in regular mode and in Learning Mode. This way we make sure the quiz behaves the same in regular mode and in learning mode. And we don't need to worry about two different cases when we extend quiz functionality with hooks.

The only difference would be is that in Learning Mode the hooks would swallow the buffer output to prevent attached actions from rendering custom markup during the hook execution. Because Learning Mode is all about Sensei's opinionated view on how the experience should look like and all the required markup is rendered by the Sensei Theme already.